docs/types.rst

```diff
@@ -513,16 +513,21 @@ Data location
 -------------
 
 Every complex type, i.e. *arrays* and *structs*, has an additional
-annotation, the "data location", about whether it is stored in memory or in storage. Depending on the
-context, there is always a default, but it can be overridden by appending
-either ``storage`` or ``memory`` to the type. The default for function parameters (including return parameters) is ``memory``, the default for local variables is ``storage`` and the location is forced
-to ``storage`` for state variables (obviously).
+annotation, the "data location", about whether it is stored in memory or in storage.
+All complex types must have an explicit data location, except for constants,
+event parameters (both default to memory), state variables (default to storage),
+and external function parameters (discussed below).
```

Comment:
Please change this to say that there are three data locations: memory, storage and calldata. Calldata is only valid for parameters of external contract functions and also has to be used there.